core/src/main/scala/fs2/job/JobManager.scala

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package job
+
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import cats.effect.{Concurrent, Timer}
+
+import fs2.concurrent.{Queue, SignallingRef}
+
+import scala.{Array, Boolean, Int, List, Option, None, Nothing, Some, Unit}
+import scala.collection.JavaConverters._
+import scala.util.{Left, Right}
+
+import java.lang.{RuntimeException, SuppressWarnings}
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * A coordination mechanism for parallel job management. This
+ * structure provides mechanisms for aggregated notifications
+ * from jobs, deterministic cancelation, "fire and forget"
+ * submission, and sidecar-style cancelation of direct streams.
+ */
+final class JobManager[F[_]: Concurrent: Timer, I, N] private (
+    notificationsQ: Queue[F, Option[N]],
+    dispatchQ: Queue[F, Stream[F, Nothing]]) {
+
+  import JobManager._
+
+  private[this] val meta: ConcurrentHashMap[I, Context[F]] = new ConcurrentHashMap[I, Context[F]]
+
+  val notifications: Stream[F, N] = notificationsQ.dequeue.unNoneTerminate
+
+  /**
+   * Submits a job for parallel execution at the earliest possible moment.
+   * If the job submission queue is full, this function will asynchronously
+   * block until space is available. Note that a job is visible as Pending
+   * immediately upon submission, even when space is unavailable.
+   *
+   * Attempting to submit a job with the same id as a pre-existing job will
+   * produce false.
+   */
+  @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments", "org.wartremover.warts.Equals"))
+  def submit[R](job: Job[F, I, N, R]): F[Boolean] = {
+    val run = job.run
+      .map(_.swap.toOption)
+      .unNone
+      .map(Some(_))
+      .evalMap(notificationsQ.enqueue1)
```

Comment:
Well, really the only other possibility is dropping notifications. We could do that by making `notificationsQ` a `SignallingRef` rather than a `Queue`, but that seems like even more surprising.

Reply:
Hm, I was thinking of the case where notifications represent something like "percent complete" and having my job stall because I wasn't consuming them wouldn't be desirable.

Reply:
Hmm. "Percent complete" would definitely be much more of a `Signal`-style notification. I wonder if we need both types?